Add time-to-first-position metric to `maker`

luckysori opened the pull request:
Fixes https://github.com/itchysats/itchysats/issues/1895.

Since this metric is obviously only relevant for a maker, I was inspired to attempt to start extracting code that is only used by the `maker` binary into the `maker` crate. These changes aren't strictly necessary, but I think they are in the right direction.

luckysori committed:
Move maker-specific actors from daemon to maker

luckysori committed:
Add necessary feature to bitmex-stream

Running `cargo hack check` revealed that `bitmex-stream` does not
build by itself. I thought that we were already verifying this by
running `clippy` separately on all our crates, so this is a surprise.

luckysori committed:
Remove unused dependencies

Identified by running `cargo udeps`.

luckysori committed:
Record time-to-first-position metric in maker DB

thomaseizinger commented on future-timeout/Cargo.toml:
```diff
@@ -0,0 +1,7 @@
+[package]
+name = "future-timeout"
```
Smallest crate in the universe :D

Perhaps just duplicate this?

luckysori replied:
Hahaha fair enough.

thomaseizinger commented on maker/src/cfd.rs:
```diff
@@ -0,0 +1,711 @@
+use crate::collab_settlement;
```
I wonder why this isn't marked as moved?

luckysori replied:
That is weird! Maybe the original file wasn't deleted?

thomaseizinger commented on daemon/src/rollover_maker.rs:
```diff
@@ -0,0 +1 @@
+
```
What is this? :)

luckysori replied:
Classic! I'll see how many files were not moved properly.

thomaseizinger commented on daemon/src/db/time_to_first_position.rs:
```diff
@@ -0,0 +1,262 @@
+    let row = sqlx::query!(
+        r#"
+        SELECT
+            time_to_first_position.first_seen_timestamp
```
```suggestion
            first_seen_timestamp
```

I don't think you need the table name here!

da-kami replied:
yeah, in statements that only involve one table they are somewhat of an overkill.

luckysori replied:
Yep, can do. I think we were trying to be consistent with always referencing the table. But it is overkill.

thomaseizinger commented on daemon/src/db/time_to_first_position.rs:
```diff
@@ -0,0 +1,262 @@
+    let row = sqlx::query!(
+        r#"
+            SELECT
+                time_to_first_position.first_position_timestamp
```
Same here.

thomaseizinger commented on daemon/src/db/time_to_first_position.rs:
```diff
@@ -0,0 +1,262 @@
+
+        if !have_seen_before(&mut conn, taker_id).await? {
+            insert_first_seen(&mut conn, taker_id, timestamp).await?;
+        }
```
Could potentially replace this with SQL?

https://stackoverflow.com/a/19337206

bonomat replied:
nice idea. 👍

da-kami replied:
I think this should work for this instance, but might be a good idea to have another look yourself on how sqlite's conflict resolution algorithm works when using `REPLACE`:

https://sqlite.org/lang_conflict.html

You will definitely need a `UNIQUE` constraint on the `taker_id` then, otherwise `REPLACE` might not work as expected.
Given that we can have `NULL` values in the table please add tests that show that this works as intended!

luckysori replied:
I like this proposal, but `sqlx` doesn't accept the query I've written. Perhaps someone can help me with this.

bonomat replied:
Happy to help out. I'm the master of SQL and killed once a production DB at Siemens using recursive read queries!!!

bonomat replied:
This was bogus...

bonomat replied:
Hold on, this does not work, you might accidentally update the wrong taker's fields 😬

bonomat replied:
So, for inserting the `first_seen_timestamp` you can do this: 

```sql
INSERT OR IGNORE INTO time_to_first_position (taker_id, first_seen_timestamp)
select 'test4', 4
```

It will only insert if it does not exist yet based on the unique key `taker_id`

For updating the first open position you can do: 

```sql
UPDATE time_to_first_position
SET first_position_timestamp = 42
WHERE taker_id = 'test' and first_position_timestamp is null;
```

luckysori replied:
Thanks for giving it a go! I'll try to add it to this PR.

luckysori replied:
Why can't we do the same thing for the `first_position_timestamp`? The approach you propose for the `first_seen_timestamp` seems to work (i.e. tests pass).

bonomat replied:
Afaik, if `taker_id` is already present then `insert or ignore` will ignore it and won't do anything.

luckysori replied:
> Afaik, if `taker_id` is already present then `insert or ignore` will ignore it and won't do anything.

Confirmed!

bonomat commented on daemon/Cargo.toml:
```diff
@@ -42,7 +43,6 @@ sqlx = { version = "0.5", features = ["offline", "sqlite", "uuid", "runtime-toki
 statrs = "0.15"
 thiserror = "1"
 time = { version = "0.3", features = ["serde", "macros", "parsing", "formatting", "serde-well-known"] }
-future-timeout = { path = "../future-timeout" }
```
did you check if we can remove more dependencies now?

bonomat replied:
Saw it happening a few commits later :)

bonomat commented on daemon/migrations/20220502172229_create_time_to_first_position_table.sql:
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE IF NOT EXISTS time_to_first_position (
+    id integer PRIMARY KEY autoincrement,
+    taker_id text NOT NULL,
```
Imho this should be `UNIQUE`

thomaseizinger replied:
There is a unique constraint further down!

da-kami replied:
Is there a specific reason / advantage for (only) creating the `UNIQUE` constraint implicitly through adding an index? 
I agree with @bonomat that it would be easier to understand if we'd have (an additional) `UNIQUE` constraint here.

luckysori replied:
I probably just forgot the `UNIQUE` label!

bonomat commented on daemon/src/db/time_to_first_position.rs:
```diff
@@ -0,0 +1,262 @@
+        Some(timestamp) => timestamp,
+    };
+
+    let timestamp = OffsetDateTime::from_unix_timestamp(timestamp)?;
```
Might be a usecase for `transpose` https://doc.rust-lang.org/std/option/enum.Option.html#method.transpose :) 

Maybe like this: 

```
timestamp
  .map(|row| OffsetDateTime::from_unix_timestamp(row.first_seen_timestamp))
  .transpose()?
```

luckysori replied:
I tried it originally, but it was not very readable at all haha, because it's actually a `Result<Option<Option<OffsetDateTime>>` and I don't think you can pattern match much since `sqlx::query!` outputs "an ad-hoc anonymous struct type".

But do give it a go and see if you can produce something better!

bonomat replied:
Pushed one commit just for fun :) https://github.com/itchysats/itchysats/pull/1948/commits/a876482ac26d4dd106dec23a8a61e3b7b96059d2

da-kami reviewed (commented):
ack metrics concept! 
I like where the extracing code part is going, but it feels almost a bit much for this PR :)

luckysori committed:
Allow sqlx queries to be defined under cfg(test)

For that we have to modify the `prepare_db.sh` script. Otherwise,
`sqlx` queries defined under the `test` feature flag will not be
prepared when we run the script.

luckysori committed:
Derive Copy for CfdNewOfferParamsRequest

bonomat commented on daemon/src/db/time_to_first_position.rs:
```diff
@@ -0,0 +1,215 @@
+        let row = sqlx::query!(
+            r#"
+            SELECT
+                time_to_first_position.first_seen_timestamp
```
🙃
You could remove `time_to_first_position` here

bonomat committed:
Use `map` and `transpose` for fun :)

bonomat reviewed (approved):
I think this is good to go!

luckysori commented on daemon/src/db/time_to_first_position.rs:
```diff
@@ -159,19 +159,15 @@ mod tests {
-        let timestamp = OffsetDateTime::from_unix_timestamp(timestamp)?;
-
-        Ok(Some(timestamp))
+        let timestamp = row
+            .map(|row| {
+                row.first_seen_timestamp
+                    .map(OffsetDateTime::from_unix_timestamp)
+            })
+            .unwrap_or(None)
+            .transpose()?;
+
+        Ok(timestamp)
```
Amazing! Thank you 😍
